feat(kubernetes): Add support for excluding event loading on manifest…

mtweten opened the pull request:
…s retrieval

This change allows optionally skipping kubernetes event retrieval when retrieving manifests. 

When live manifest mode is enabled, scenarios that result in a lot of manifests being loaded will put a lot of extra pressure on the kubernetes API server due to event loading. 

Specifically, we ran into this scenario when we had a lot of deployment pipelines running concurrently, and each deployment involved a decent number of manifests (~30). In Orca, the WaitForManifestStableTask (when live manifest mode is enabled) iterates through the manifests involved in the deployment and calls the retrieve manifest API for each one, which causes a lot of full event loads to occur, which tanks (in our case) the kubernetes API server. However, WaitForManifestStableTask does not actually need any event information to do its job, so in that particular case we can skip loading events entirely.

See https://github.com/spinnaker/orca/pull/2997 for the corresponding Orca change.

ezimanyi commented on clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/model/ManifestProvider.java:
```diff
@@ -27,6 +27,17 @@
   List<T> getClusterAndSortAscending(
       String account, String location, String kind, String app, String cluster, Sort sort);
+
+  List<T> getClusterAndSortAscending(
```
This is only ever called in one place (`getDynamicManifestFromCluster`); the only existing call to that REST endpoint is from `DynamicResolveManifestTask` in `orca` which also doesn't use the events, so I think we can safely just always omit events from this call.  If someone adds a use case that needs the events, they can add in the flag and pass it through.  (Notably this is never exposed through gate, only internal REST endpoints which we don't guarantee backwards compatibility for.)

ezimanyi commented on clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/model/ManifestProvider.java:
```diff
@@ -27,6 +27,17 @@
 
   T getManifest(String account, String location, String name);
 
+  T getManifest(String account, String location, String name, boolean includeEvents);
```
Given that this is only called in 2 places, it might be simpler to just always require the boolean flag.  The only remaining location that would call the old version is `KubernetesV2JobProvider` which also doesn't need events so should just pass in `false`.

ezimanyi commented on clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/caching/view/provider/KubernetesV2LiveManifestProvider.java:
```diff
@@ -84,7 +90,10 @@ public KubernetesV2Manifest getManifest(String account, String location, String
 
-    List<KubernetesManifest> events = credentials.eventsFor(kind, namespace, parsedName.getRight());
+    List<KubernetesManifest> events =
+        !includeEvents
```
This would be easier to read if we checked the positive condition and inverted to:
```
includeEvents ? credentials.eventsFor(kind, namespace, parsedName.getRight()) : Collections.emptyList()
```

ezimanyi commented on clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/caching/view/provider/KubernetesV2ManifestProvider.java:
```diff
@@ -127,19 +147,24 @@ public KubernetesV2Manifest getManifest(String account, String location, String
-            .stream()
-            .map(KubernetesCacheDataConverter::getManifest)
-            .collect(Collectors.toList());
+        !includeEvents
```
Same comment as above about checking the positive condition `includeEvents` rather than `!includeEvents`

ezimanyi commented on clouddriver-web/src/main/groovy/com/netflix/spinnaker/clouddriver/controllers/ManifestController.java:
```diff
@@ -52,15 +53,24 @@ public ManifestController(List<ManifestProvider> manifestProviders, RequestQueue
+      @PathVariable String location,
+      @PathVariable String name,
+      @RequestParam(value = "includeEvents", required = false, defaultValue = "true")
+          boolean includeEvents) {
```
Looks like you forget to pass this parameter to the call to `getManifest` below?

mtweten replied:
Good catch, looks like I messed this up when rebasing some of my changes

mtweten commented:
Thanks for the review Eric!

Addressed your comments, I agree with simplifying the number of APIs - I was just initially a little worried and wanted to do it in a super passive way so I didn't inadvertently break anything.

Only thing I wanted to be sure of was the change in `getDynamicManifestFromCluster` - essentially we're just saying the contract for that method is that events aren't returned, and if that functionality is eventually needed it can be added in the future?
